### PR TITLE
Add filtering ability for all backend API ListXXX requests

### PR DESCRIPTION
**Pull request by neuromage**
This PR adds filtering capability to all ListXXX requests in the backend API. As the context for filtering needs to be preserved with that of sorting, I've unified both under the new 'list' package, which attempts to streamline the previous logic for sorting and generating next page tokens. High-level overview of what's in this PR:

* Introduce the `Filter` protocol buffer which is used to transmit filtering requests from the frontend to the backend.
* The package `filter` handles parsing the Filter proto and is able to apply corresponding WHERE clauses when constructing SQL statements.
* The package `list` handles both sorting and filtering, and is used by all ListXXX requests down the stack. It generates next page tokens to ensure both sorting and filtering criteria carries over to the next page of results.
* All models implement list.Listable, which defines a minimal interface that needs to be satisfied by any model that supports list operations.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/537)

### CODE REVIEW

File: backend/api/filter.proto [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    // |long_value|, |string_value| or |timestamp_value|.
+    EQUALS = 1;
+    NOT_EQUALS = 2;
+    GREATER_THAN = 3;
+    GREATER_THAN_EQUALS = 5;
+    LESS_THAN = 6;
+    LESS_THAN_EQUALS = 7;
+
+    // Checks if the value is a member of a given array, which should be one of
+    // |int_values|, |long_values| or |string_values|.
```
[the remark this answers is not in the document]
Comment (neuromage) [reply backend/api/filter.proto]: There's no good reason other than I thought it unlikely we'd want to query specific dates vs a range of dates. If we ever want to do so, we can add it easily.

File: backend/api/filter.proto [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+//     key: "label"
+//     op: IN
+//     string_values {
+//       value: 'label_1'
+//       value: 'label_2'
+//     }
+//   }
+// }
+message Filter {
+  // All predicates are AND-ed when this filter is applied.
```
[the remark this answers is not in the document]
Comment (neuromage) [reply backend/api/filter.proto]: The logic of the filter is part of the contract, so I think having it here is useful for anyone trying to create a Filter proto.

File: backend/api/filter.proto [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    // Checks if the value is a member of a given array, which should be one of
+    // |int_values|, |long_values| or |string_values|.
+    IN = 8;
+  }
+  Op op = 1;
+
+  string key = 2;
+  oneof value {
+    int32 int_value = 3;
```
[the remark this answers is not in the document]
Comment (neuromage) [reply backend/api/filter.proto]: I don't think so. The client can be in any language that supports protos, like Go. In which case types do matter.

**Comment by neuromage** [comment]
@yebrahim I can send out another PR with changes if you'd like me to make any. Please let me know. Thanks!

**Comment by yebrahim** [comment]
I don't have very strong opinions on the comments here, so I'm fine with these resolutions.
Thanks for making the change!

**Pull request merged**